misk/src/main/kotlin/misk/web/exceptions/ActionExceptionMapper.kt

```diff
@@ -0,0 +1,31 @@
+package misk.web.exceptions
+
+import misk.exceptions.ActionException
+import misk.web.Response
+import misk.web.ResponseBody
+import misk.web.marshal.StringResponseBody
+import misk.web.mediatype.MediaTypes
+import okhttp3.Headers
+
+/**
+ * Maps [ActionException]s into the appropriate status code. [ActionException]s corresponding
+ * to client-errors (bad requests, resource not found, etc) are returned with full messages
+ * allowing the client to deterine what went wrong; exceptions representing server errors
+ * are returned with just a status code and minimal messaging, to avoid leaking internal
+ * implementation details and possible vulnerabilities
+ */
+internal class ActionExceptionMapper : ExceptionMapper<ActionException> {
+    override fun toResponse(th: ActionException): Response<ResponseBody> {
+        val message = if (th.statusCode.isClientError) th.message ?: th.statusCode.name
+        else th.statusCode.name
+        return Response(StringResponseBody(message), HEADERS, statusCode = th.statusCode.code)
+    }
+
+    override fun canHandle(th: Throwable): Boolean = th is ActionException
+
+    private companion object {
+        val HEADERS: Headers =
+                Headers.of(listOf("Content-Type" to MediaTypes.TEXT_PLAIN_UTF8).toMap())
```

Comment:
I think returning errors as plaintext is nice.

Reply:
For default errors yeah. I can see a case for returning structured errors for more complex scenarios (e.g. giving a high level status code + sub-code providing more details, or returning the specific list of problematic fields when throwing a validation exception). Could argue those cases should be handled by returning specific response types rather than exceptions, but often they are actually exceptional and forcing the path to go through response types results in loss of type information in the actions (e.g. need to return generic ``Response`` objects rather than the "expected" response type)